File: Swift/ParallelChange/ParallelChangeTest/AuthenticationService.swift
import XCTest

class AuthenticationService {

    //
    // the goal is to replace the method above with this one:
    // func isAuthenticated(role, userId)
    //

    func isAuthenticated(userId: Int) -> Bool {
        return userId == 12345
    }
}

class AuthenticationClient {
    let authenticationService: AuthenticationService

    init(authenticationService: AuthenticationService) {
        self.authenticationService = authenticationService
    }

    func run() {
        let authenticated = authenticationService.isAuthenticated(33)
        print("authenticated: \(authenticated)")
    }
}

class YetAnotherClient {
    func run() {
        AuthenticationService().isAuthenticated(100)
    }
}

class AuthenticationServiceMain: XCTestCase {
    func runExample() {
        let client = AuthenticationClient(authenticationService: AuthenticationService())
        client.run()
    }

    func testRunExample() {
        runExample()
    }
}

File: Ruby/authentication_service.rb
class AuthenticationService

  #
  # the goal is to replace the following method with:
  # def authenticated?(role, user_id)
  #

  def authenticated?(user_id)
    user_id == 12345
  end
end

class AuthenticationClient
  def initialize(authentication_service)
    @authentication_service = authentication_service
  end

  def run
    authenticated = @authentication_service.authenticated?(33)
    puts "authenticated: #{authenticated}"
  end
end

class YetAnotherClient
  def run
    AuthenticationService.new.authenticated?(100)
  end
end

if __FILE__ == $PROGRAM_NAME
  client = AuthenticationClient.new(AuthenticationService.new)
  client.run
end
